Use an archive if possible when fetching fails and --airgap is provided

clintoncwolfe opened the pull request:
<!--- Provide a short summary of your changes in the Title above -->

## Description

Adds a new mode to the `inspec archive` command. Under the following conditions:

 * When the `--airgap` option is provided
 * and the top-level profile has one or more local dependencies on an archive (a zip file or tarball)
 * and fetching any other dependency in the tree fails for any of the local, git, or URL fetchers,

Instead of failing the operation, use the copy of the unreachable profile that is contained in the archive.

### Implementation Notes

This works by fooling the cache in a couple of ways. During an `archive` operation, the vendor cache is normally purged. However, if `--airgap` is provided, the cache is warmed by locating and extracting any archives that are at the top-level of the profile tree.

Next, when profiles fail to be fetched, it is usually because their cache key is dependent on information only available on the network. However, that same information is available in the lockfile of the archive. So, find the lockfile, find the cache key, and update the fetcher to make it believe that it already knows its cache key (which we know will be a "hit").

### Further Work

This PR currently provides implementations for the `local`, `url`, and `git` fetchers.  Currently no support is provided for the `supermarket` ~~or `compliance`~~ fetchers. UPDATE: the compliance fetcher is also included.

It's also unclear if `--airgap` is the right word for this. One idea is to use that same option on any command that is able to operate offline.  Is `--offline` a better word?

## Related Issue

Closes #4727

Refs #4519 

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New content (non-breaking change)
- [ ] Breaking change (a content change which would break existing functionality or processes)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [x] I have read the **CONTRIBUTING** document.

clintoncwolfe committed:
Add .invalid  TLD to test hostanmes

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Remove constructors from raise statements

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Functional test and fixtures for git fetcher fallback

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
When archiving with --airgap, warm the vendor cache from any available local archives

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Add an optional API method to update the internal state of a fetcher

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Alter resolver to fallback to an archive if in airgap mode

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Test fixture for url fetcher

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Implementation of archive fallback for the URL fetcher

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Test fixtures and functional test for local fetcher

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Implementation for local fetcher

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Test fixture and functional test for compliance fetcher

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
A terrible implementation of fetcher fallback for compliance fetcher

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

miah commented on lib/inspec/fetcher/git.rb:
```diff
@@ -112,6 +112,19 @@ def resolved_source
+      end
+      changed
+    end
+
```
```ruby
%i{branch tag ref}.map { |opt_name| update_ivar_from_opt(opt_name, opts) }.any?
```

clintoncwolfe replied:
At first I grumbled (another @#$!% style opinion) but then I scrolled over and saw the `.any?` at the end and quite liked it. Thanks!

zenspider replied:
Maybe you could not bias towards this attitude to begin with?

The code in `update_from_opts` is not clear:

* The false literal is a smell.
* The `var = x; collection.each { ... }; var` pattern is a smell and almost always a sign that Enumerable isn't being used to our advantage.
* The logical `||=` requires thinking about.

Whereas `collection.map { ... }.any?` expressly conveys what this method is trying to accomplish.

miah commented on lib/inspec/fetcher/local.rb:
```diff
@@ -100,6 +100,19 @@ def sha256
+        changed = true
+      end
+
+      changed
```
```ruby
changed = @archive_shasum != opts[:sha256]
@archive_shasum = opts[:sha256]
changed
```

clintoncwolfe replied:
I think there are a lot of ways to write the same thing, and I will always prefer the longer, dumber, clearer way.

zenspider replied:
I might change the suggested code to the following:

```ruby
original = @archive_shasum
@archive_shasum = opts[:sha256]
original != opts[:sha256]
```

The only difference between those two versions is whether you want to express the intent of `changed` vs `original`... they're similar enough to me that I don't need to think about either of them for more than half a second.

Your original code is longer and "dumber", yes, but not clearer in any sense. It took me probably 20 seconds to fully comprehend after getting through these:

* The true/false literals are a smell.
* The `var = x; do stuff; var` pattern is a smell.
* The conditional requires thinking about.
* There's no reason NOT to assign if they're the same, so why the conditional?

clintoncwolfe replied:
We should talk offline about subjective review.

miah commented on lib/inspec/fetcher/url.rb:
```diff
@@ -118,6 +118,19 @@ def to_s
+        changed = true
+      end
+
+      changed
```
Identical code fetcher/local.rb#update_from_opts. Push up to `Inspec::Plugins::Fetcher`

clintoncwolfe replied:
I hate to make "do something ugly and illogical" (having the opts not match the ivar name) the superclass implementation. To me, that obliges subclasses to override, which is not what you generally want.

zenspider replied:
This doesn't make sense to me. You added the same named method to the superclass that returns `false` and thus "obliges subclasses to override". Can you clarify?

Having identical code in two sibling classes and pushing them up to the abstract superclass IS a very normal refactoring. This pushback doesn't make sense to me.

clintoncwolfe replied:
First, thanks - you're right, I forgot I did provide a default superclass implementation that does no update and returns `false`. So any any subclass would be obliged to override if they wanted any behavior that wasn't that; but at least is it is a **sensible** default.

I'll cave on this one - I'll replace the sensible superclass implementation with the weird one from url/local. Any new fetcher would need to remember to override that with a sane updater, because it's not likely they'll happen to choose the same ivar name that local and url did. 

That trades a small x2 DRY problem for reduced clarity and increased extensibility cost.

miah commented on lib/inspec/plugin/v1/plugin_types/fetcher.rb:
```diff
@@ -65,6 +65,28 @@ def cache_key
+        return false if old_val == new_val
+
+        instance_variable_set(ivar_sym, new_val)
+        true
```
Same pattern as above.

clintoncwolfe replied:
Thanks, I've put that in.

miah commented on lib/inspec/profile_vendor.rb:
```diff
@@ -51,8 +51,10 @@ def profile_opts
-    def vendor_dependencies
+    def vendor_dependencies(opts)
+      # This deletes any existing vendor/ directory
       delete_vendored_data
```
Should this be a bang! method?

clintoncwolfe replied:
Sure

miah commented on lib/inspec/profile_vendor.rb:
```diff
@@ -80,5 +82,40 @@ def extract_archives
+
+        # Determine if it is an archive
+        file_provider = FileProvider.for_path(dep_profile_path)
+        next unless file_provider.is_a?(ZipProvider) || file_provider.is_a?(TarProvider)
```
If the fetcher system is plugable, then we have to update this list dynamically.

zenspider replied:
Agreed. Typechecks are a smell.

Both `ZipProvider` and `TarProvider` subclass from `FileProvider`... If they moved to subclassing `ArchiveProvider` (which subclassed `FileProvider`) then only one check needs to be made here and it is more resilient to change.

clintoncwolfe replied:
This is the `FileProvider` system, not the `Fetcher`, as I understand it, and is used in other places to detect whether a file is an archive - in fact, just a few lines above at https://github.com/inspec/inspec/blob/8b5789004ac31f4b6e7210421b119af41f483e3e/lib/inspec/profile_vendor.rb#L74 . Good idea or not, "how do I detect an archive in InSpec" has at least one existing answer - and it is based on typechecking. Sounds like you'd like a  different approach, and you're like me to alter an unrelated class hierarchy to avoid a code smell.

Respectfully, I feel that is outside the scope of this PR. I'll open an issue. #4824

miah commented on lib/inspec/profile_vendor.rb:
```diff
@@ -80,5 +82,40 @@ def extract_archives
+        # The archive (probably) contained a vendor cache of its own.
+        # We must flatten that, so that all vendor cache entries are top-level.
+        Dir["#{dep_profile_cache_dir}/vendor/*"].each do |sub_dep_cache_dir|
+          FileUtils.mv(sub_dep_cache_dir, cache_path)
```
Collision if two dependencies have the same name and get moved to cache_path

clintoncwolfe replied:
That's just going to be a limitation. We'll need to document that.

zenspider replied:
doco doco? or comment here?

miah commented on test/functional/fetchers_test.rb:
```diff
@@ -1,16 +1,16 @@
   # Refs #4726
   describe "when fetchers fetch a bad dependency" do
-    let(:fetcher_profiles) { "#{profile_path}/fetcher-failures" }
+    parallelize_me!
```
We can't parallelize this whole suite?

clintoncwolfe replied:
I believe we can, yes. That was a holdover from early work.

miah commented on test/functional/fetchers_test.rb:
```diff
@@ -54,4 +54,52 @@ def assert_fetcher_failed_cleanly(run_result, error_regex, profile_location)
+
+    def teardown
+      FileUtils.rm_rf "#{fetcher_profiles}/#{profile_name}/vendor"
+      FileUtils.rm_f "#{profile_name}-0.1.0.tar.gz"
```
*scary* we need to be really careful with rm_rf and rm_f especially since we're not using temp directories anywhere here (we should though).

clintoncwolfe replied:
Agreed

zenspider commented on lib/inspec/fetcher/url.rb:
```diff
@@ -127,8 +140,10 @@ def parse_uri(target)
+
       file = @archive_path || temp_archive_path
-      OpenSSL::Digest::SHA256.digest(File.read(file)).unpack("H*")[0]
+      @archive_shasum = OpenSSL::Digest::SHA256.digest(File.read(file)).unpack("H*")[0]
```
This is cleaner (and gets us closer to `-w`) if it is in `||=`-style.

zenspider commented on test/functional/fetchers_test.rb:
```diff
@@ -54,4 +54,52 @@ def assert_fetcher_failed_cleanly(run_result, error_regex, profile_location)
+      _(run_result.stderr).must_be_empty
+      _(run_result.stdout).must_include "Finished archive generation"
+      assert_exit_code(0, run_result)
+    end
```
Please use `prepare_examples` or the `tmpdir` option in `run_inspec_process`.

clintoncwolfe replied:
Agreed, I'll aim to rework the tests to use tmpdir option to `run_inspec_process`.

clintoncwolfe committed:
bangify vendor_dependencies

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
make a one-liner per miah

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
parallelize fetcher tests, conditionalize automate fetcher

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Push up update_from_opts implementation

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Switch to using tmpdir option during testing to avoid rm calls

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Adjust usage message for --airgap option

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Linting

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

zenspider committed:
Fix test failure for call to `vendor!` w/o options.

Had to make options available for the mocking as well.

Signed-off-by: Ryan Davis <zenspider@chef.io>

zenspider committed:
Fixed resolver_test.rb by making Inspec::Config.cached a lazy accessor.

Signed-off-by: Ryan Davis <zenspider@chef.io>

james-stocks commented on lib/inspec/cli.rb:
```diff
@@ -173,6 +173,8 @@ def vendor(path = nil)
   option :overwrite, type: :boolean, default: false,
     desc: "Overwrite existing archive."
+  option :airgap, type: :string, default: false, banner: "", # Actually a boolean, but this suppresses the creation of a --no-airgap...
+    desc: "Prefer using existing local archives to fetching profiles."
```
Reading the PR description, it sounds like this option allows a fallback to local files when fetching fails; rather than preferring local files? So maybe `"Allow InSpec to use local archives when fetching dependencies"` ?

The term "local archives" could maybe be more explicitly written as "existing content in the vendor subdirectory"?

clintoncwolfe replied:
The behavior is to attempt to fetch first, then fallback (ie, reprogram the fetcher to point at the archive) and re-fetch. If the second fetch fails, give up. 

`Fallback to use local archives if fetching fails`?

Or should there be a `--airgap no-fetch` (or similar) value indicating no first attempt should be made?

james-stocks replied:
I was agreeing with the functionality and just critiquing the `desc:` string. It uses the word "prefer" which implies it will fetch local first; which doesn't reflect the functionality.
I'm just suggesting changing this string.

At the same time, it would be awesome to add an option to indicate that InSpec should immediately use the local content first, and not attempt a remote fetch.

e.g.

```
--local-dependencies never
--local-dependencies allow
--local-dependencies always
```

clintoncwolfe replied:
OK. Let's leave the functionality alone for now then, so we can see how customers like the feature as-is. Let's revise the usage string to clarify that it is a fallback operation.

james-stocks commented on lib/inspec/dependencies/resolver.rb:
```diff
@@ -82,5 +92,41 @@ def resolve(deps, top_level = true, seen_items = {}, path_string = "") # rubocop
       graph
     end
+
+    def fallback_to_archive_on_fetch_failure(dep)
```
Method name like `can_fallback_to_archive?` might be more expressive for what this method does

james-stocks replied:
My suggestion there is wrong because this method makes some changes to the current state, it doesn't just return a boolean.

james-stocks replied:
I would disagree with "very narrow use case" because it's the main use case for enterprise-scale customers (wrapper profile in an air gapped environment).
Maybe delete the current description paragraph and give 1-2 sentences that state what the method does and what the return condition is.

clintoncwolfe replied:
Hrm, as written, that does read as editorializing. I'll revise. I don't mean that it is uncommon or unimportant, just that it is a very specific situation - a very special set of conditions have to be met in order for this feature to be able to work.

james-stocks commented on test/fixtures/profiles/fetcher-failures/local-dep-on-bad-auto-archive/inspec.yml:
```diff
@@ -0,0 +1,8 @@
+  platform: os
+depends:
+  - name: auto-dep-on-missing
+    path: ../auto-dep-on-missing-0.1.0.tar.gz
```
👍 this appears to match the customer case - the tar.gz is available at the given path, but the dependency defined _inside_ that  .tar.gz is unreachable. Awesom!

clintoncwolfe committed:
A clumsy but effective way of usually dodging windows filesystem length restrictions.

Still definitely possible to run afoul of that issue, but that is not specific to
this feature, and frankly one might consider increasing thier FS path length
as a solution, as well.

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Accept a refactornig suggestion

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Compress some code

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Clarify comment on fallback_to_archive_fetch_failure

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

clintoncwolfe committed:
Clarify usage message for --airgap option

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>

miah commented on lib/inspec/plugin/v1/plugin_types/fetcher.rb:
```diff
@@ -65,6 +65,33 @@ def cache_key
+      # can be updated with information that might lead to a successful
+      # retrieval of alternative content, this method may be called.
+      #
+      #  Default implementation makes a peculiar assumption that the class has
```
nitpick; extra space =)

miah reviewed (approved):
Thanks!

zenspider commented on lib/inspec/profile_vendor.rb:
```diff
@@ -107,14 +107,20 @@ def warm_vendor_cache_from_archives
-        file_provider.extract(dep_profile_cache_dir)
+        # Use a shorter name here in hopes of dodging windows filesystems path length restrictions
+        actual_dep_profile_cache_dir = "#{cache_path}/#{fetcher.cache_key}"
+        short_dep_profile_cache_dir = "#{cache_path}/#{SecureRandom.hex[0..6]}"
```
This doesn't have to be a random name, does it? Literally `"#{cache_path}/short"` would work. If the concern is collision, then mktmpdir might be a better bet.

clintoncwolfe committed:
PR feedback re: cache warming

Signed-off-by: Clinton Wolfe <clintoncwolfe@gmail.com>
